### DISCUSSION TITLE
Bundled Txs = 80% Lower Fees

### DISCUSSION THREAD

**Discussion by Jpja** [CounterpartyXCP/CIPs#127] [CIPs]
By bundling multiple Counterparty messages from several users into one Bitcoin transaction, I believe it can be possible to reduce fees by >80%.

Key assumptions:

- A bitcoin address can be recovered from its 32 byte ECDSA signature.
- Counterparty introduces witness script encoding, which receives a 75% block discount.

**Anatomy of a Counterparty Transaction**

A typical Counterparty transaction looks like [this](https://jpja.github.io/Electrum-Counterparty/decode_tx?tx=1fb3036735a07c4de97459dd06de7ffe5a8bc5fbcbfb9f3f08376aaea1ee6b43). One input, one change output, and an op_return message. The message begins with `CNTRPRTY` and tells Counterparty to perform one specific instruction. The message takes up something like 50-70 bytes, while the overall Bitcoin tx is ~250 Bytes. If bech32 is used the transaction gets a discount  and ends up at around 150 vBytes.

**Compressing the Message**

The message uses a lot of fixed width fields. This makes a lot of empty bytes. and the msg can easily be compressed. For example, the "send amount" takes up 8 bytes, so to send one token you'll engrave `0000000000000001` on the blockchain. I tested out a [simple algorithm](https://jpjanssen.com/compressed-xcp-transactions/). The result varied a lot between message types, ranging from 6 to 67%:

![image](https://github.com/CounterpartyXCP/cips/assets/8447505/4c804435-fe25-4f6b-99fb-6020c1070c16)

Important note is that the absolute saving is 7 to 28 bytes, so the benefit on the overall Bitcoin tx size is limited.

 **Bundling Messages from One Sender**

Now the logic is one CNTRPRTY message, one Bitcoin transaction. I don't see any good reason why we cannot put N messages into one transaction. Think of a message as a line of code. Bundling will make several lines that Counterparty executes one after another. 

An example is when I helped with issuing Monster Book cards. This took one hundred chained Bitcoin transactions. A big hassle! I wish I could encode all one hundred instructions in one Bitcoin tx. Also everyday tasks like issuing a token and then putting it on dispenser would be great to do inside one Bitcoin tx.

However useful, the average daily user who only makes one action would not see any benefit. Therefore we should consider:

**Bundling Messages from Multiple Senders**

One user, call him the broadcaster, will add CNTRPRTY messages from multiple users into a Bitcoin transaction. This adds complexity but perhaps it's worth it?

Obviously each user must sign their own message. I believe a 32 byte ECDSA signature can do. If the message is 50 bytes, this then makes the _signed message_ a total of 82 bytes. For broadcasting of multiple messages to work, some additional parameters are needed:

- nonce - 2 bytes
- xcp fee to broadcaster - 4 bytes
- last block to include 3 bytes

The messages must be serialized somehow, so say it takes in total 96 bytes. If the assumption of 75% discount is true, then you only pay for **24 bytes**! That's more than 80% cheaper than how it's currently done!

For additional messages from the same users, the signature and parameters are not repeated. This makes additional messages about ~90% cheaper.

**Encoding**

Op_return is max 80 bytes per standard rules. But protocol rules do not enforce a limit, so we could convince a miner to allow accept such transactions. If that miner has 10% hashrate, the expected wait would be 10x normal, ie 100 minutes.

Multisig - very inefficient. Not sure how much saving it would give, if any.

Segwit encoding - it has a big overhead but near 100% efficiency, if I am not mistaken. Does it also give witness discount?

Witness data - what ordinal inscription uses, and the cheapest afaik. Counterparty does not support it yet.

**Comment by jdogresorg** [comment]
Interesting idea. Reminds me of the solution Devon Weller came up with in 2017 called "Multiparty Counterparty Aggregate Transaction (MCAT)"

Unfortunately, Devon Weller's MCAT pull request has sat open for 6+ years waiting for someone to review and merge. I took the liberty of adding Devon's MCAT CIP to the repo (CIP13), and updated the discussion link to point to the forums discussion link.

https://github.com/CounterpartyXCP/cips/blob/master/cip-0013.md

**Reply by Jpja** [reply]
Good find and thanks for adding that CIP! It confirms that EDCDS signatures work how I remembered it:

> ECDSASignature is an ECDSA sep256k1 DER-encoded signature covering a sha256 hash of the entire transactionSequence.

> Reconstructing the sender's bitcoin address
The sender's address is not explicitly included in the transaction. ECDSA signatures have the property that allows recovery of the public key from the hash of the message and the signature. By using this property we can recover the public key and therefore reconstruct the sender's address.

While MCAT is a specific method for sends, what I have in mind is a general method for packing all kinds of Counterparty messages in one transaction.

Devon made a very good writeup. I will borrow his terminology and definitions for the upcoming CIP.

**Comment by Jpja** [comment]
Let us call bundled txs **MCAM** since this is very similar to the MCAT proposal. The T in MCAT is for _"send **T**ransaction"_. MCAM, in contrast, is general and works for all _counterparty **M**essages_.

Here's how an MCAM tx could be constructed:

**Server announcement**

This happens offchain; the **Server** announces intent to generate an MCAM transaction.

* First input coin, eg `fcca0631014886a909d072b5328dd3b8512780d7f4f4e7b3ce650fd8863a7172:7`
* Fee token, eg XCP (`0000000000000001`)
* Fee per byte, eg 0.01 {`00000000000186A0`)
* Expiration, eg 826,832 (`000C9DD0`)

**Message sequence A**

"User A" wants to set up a dispenser. Counterparty message: `434e5452505254590c000000000002926e0000000000000001000000000000000100000000017d784000`

Trimmed message without CNTRPTY prefix (`434e545250525459`):
`0c000000000002926e0000000000000001000000000000000100000000017d784000`

Trimmed message after compression:
`0515371714410c02926e0101017d7840`

"User A" only wants to make this one transaction. To make the message sequence, prefix the compressed message with 1 (`01`) to indicate there's one message and 16 (`10`) to signal the length of the message:
`01|10|0515371714410c02926e0101017d7840`

Now we need to sign the message sequence. Serialize the Server's parameters followed by the _message sequence:
`fcca0631014886a909d072b5328dd3b8512780d7f4f4e7b3ce650fd8863a7172|07|0000000000000001|00000000000186A|0000C9DD0|01100515371714410c02926e0101017d7840`

The 32 byte ECDSA signature signature:
`2884b7bbd70d02df70007e86ebe67f697389122a13bfa4f1ce788b2b4f9f5fd7`

Append the message sequence with a signature prefix (to determine address type) and the signature:
SSM(A) = `01|10|0515371714410c02926e0101017d7840|00|2884b7bbd70d02df70007e86ebe67f697389122a13bfa4f1ce788b2b4f9f5fd7`
 
**Message sequence B**
 
User B has two compressed messages:
`04153f09100c03ded80a`
and
`071357171541f0500c15957722ae010110c8e001d28e4bf3d70f76c6757580d4b1d9a4e9834f5cd6`

Serialized signed message:
SSM(B) = `02|0a|04153f09100c03ded80a|28|071357171541f0500c15957722ae010110c8e001d28e4bf3d70f76c6757580d4b1d9a4e9834f5cd6|80|071357171541f0500c15957722ae010110c8e001d28e4bf3d70f76c6757580d4b1d9a4e9834f5cd6`

Note that `02` indicates there are two messages, `0a` (decimal 10) is the length of the first message,  `28` (decimal 40) is the length of the second message, and `80` means it was signed with a segwit address (unlike `00` for a classic address).

**Message sequence C**

The server wants to include one message of their own. No signature is needed and `ff` is is used to indicate this:
SSM(C) = `01|19|58435001061714354431280a01ee6b2801d6b8960bebc21f80|ff`

**Constructing the MCAM message**

The MCAM message begins with the Server's parameters:
`0000000000000001|00000000000186A0|000C9DD0`
 
Compressed:
`0407153130010186A00C9DD0`

The final message that goes to the blockchain:
`434e545250525459|78|0c|0407153130010186A00C9DD0|SSM(A)|SSM(B)|SSM(C)`

Note that the input coin is not part of the message but must be the first input of the Bitcoin transaction.

`78` is the MCAM message ID.
`0c` is the length byte to the compressed parameters.
To save space, I wrote `SSM(x)` but this must of course be replaced with above serialized signed messages.

**Parsing the MCAM message**

Counterparty begins by detecting the CNTRPTRY prefix and message type `78`.

It decompresses the parameters and validates the expiration against the block height. If it is expired, the tx is invalid (and all included messages are ignored). 

Then get the first SSM and reconstruct address from signature. 

Check if address has enough funds to pay the fee. If not, ignore the entire SSM.

Process all message in the SSM in chronological order.

Repeat for all SSMs.

**Considerations**

_Fee_. The fee parameter can be set to zero. A service provider might prefer prepayments in BTC or fiat, for example, or even provide free transactions.

_Server risk_.  There will always be some time from the server's announcement to the tx is ready. The BTC fee level might have gone up with the server facing a loss. The tx may also get confirmed _after_ the expiration, in which case the server pays the full BTC fee without receiving any fees. To minimize this risk, it is likely a Server will push a tx a few blocks before the deadline, making MCAM slower than regular transactions.

_User risk_. The User needs to trust the Server making the SSM into a block. However, after expiration the SSM is invalid, so there's very limited risk.

_ECDSA_ signature. I have not looked deeper into this but I assume a byte prefix is needed to signal which address type to reconstruct (similar to how enhanced send uses it).

_Compression_. I used a home made algorithm. [See functional code](https://jpjanssen.com/compressed-xcp-transactions/). I am sure a better algorithm can be used for further encryption.

_Single party_. A User who wants to make a sequence of transactions does not _need_ an external Server. For example, when issuing an NFT collection you could generate all N issuance as one SSM and push it to the blockchain yourself.

_Gas_. Bytes per message is very low compared with traditional Counterparty txs. A discussion about gas should take place before implementing this.. Gas may be burned on top of the Server fee.

_Implementation_. This will require a protocol change. If there is demand for MCAM, the next step will be to write a formal CIP and issue a bounty for a code implementation.

***Savings***

User A's blockchain footprint would have been  ~250 Bytes (or ~150 vBytes with bech32) with a traditional Counterparty tx. 

With MCAM the SSM is 51 bytes.

Now there's also the Server's parameters which make up 22 bytes, as well as the Bitcoin tx overhead which I guess is around 150 bytes. 

With 4 messages in the MCAM tx. User A's "share" of the 172 bytes fixed cost is 43 bytes. His total footprint is 43+51 = 94 bytes. Even before taking a potential segwit discount into account, this is a clear saving.

Let us further assume a more realistic MCAM scenario including 35 messages. User A's share of the fixed bytes drops to 5 bytes, so total cost is 5+51 = 56 bytes. This already approaches a 2/3 saving over the ~150 vBytes of a regular Counterparty tx.

If Counterparty implements taproot encoding, and if I am correct that 4 bytes encoded equals 1 vbyte, then the SSM is 51/4 = 13 vbytes. The Server's parameters take up 22/4 vbytes. While assuming no discount on the Bitcoin part, the overhead is 150 + 22/4 = 156 vbytes. With 35 messages sharing the cost, User A's cost is 5 + 13 = 18 vBytes. 

Down from 150 vBytes before, the saving is 88%. Note that User A only has 1 message in SSM. The saving is significantly higher with more messages from the same user. And the assumption of 35 messages total in MCAM is just as important. More messages, more savings for each.

The Server's profit will eat up some of the savings, But even with a 100% profit margin, User A will save 76%.
